File: Backtrack/24 Game/solution.py
class Solution:
    def judgePoint24(self, cards: List[int]) -> bool:
        def dfs():
            nonlocal cards
            if len(cards) == 1:
                if abs(cards[0] - 24) < 0.01:
                    return True
                return False
            for i in range(len(cards)):
                for j in range(i+1, len(cards)):
                    a, b = cards[i], cards[j] 
                    next_nums = [a+b, a-b, b-a, a*b]
                    if b > 0.01: next_nums.append(a/b)
                    if a > 0.01: next_nums.append(b/a)
                    cards.pop(j)
                    cards.pop(i)
                    for next_num in next_nums:
                        cards.append(next_num)
                        if dfs():
                            return True
                        cards.pop()
                    cards.insert(i, a)
                    cards.insert(j, b)
            return False
        return dfs()

File: Backtrack/24 Game/solution.cpp
class Solution {
public:
    bool judgePoint24(vector<int>& cards) {
        vector<double> nums(cards.begin(), cards.end());
        auto dfs = [](this auto&& dfs, vector<double>& nums) {
            if(nums.size() == 1) {
                return abs(nums[0] - 24) < 1e-8;
            }
            for(int i = 0; i < nums.size(); ++i) {
                for(int j = i+1; j < nums.size(); ++j) {
                    vector<double> next_nums;
                    for(int k = 0; k < nums.size(); ++k) {
                        if(k != i && k != j) next_nums.push_back(nums[k]);
                    }
                    next_nums.push_back(nums[i] + nums[j]);
                    if(dfs(next_nums)) return true;
                    for(double last: {nums[i] * nums[j], nums[i] - nums[j], nums[j] - nums[i], nums[i] / nums[j], nums[j] / nums[i]}) {
                        next_nums.back() = last;
                        if(dfs(next_nums)) return true;
                    }
                }
            }
            return false;
        };
        return dfs(nums);
    }
};
